src/libraries/System.Threading/tests/SemaphoreSlimCancellationTests.cs

```diff
@@ -50,6 +50,34 @@ public static void CancelAfterWait()
             // currently we don't expose this.. but it was verified manually
         }
 
+        [Fact]
+        public static async Task Cancel_WaitAsync_ContinuationInvokedAsynchronously()
+        {
+            await Task.Run(async () => // escape xunit's SynchronizationContext
+            {
+                var cts = new CancellationTokenSource();
+                var tl = new ThreadLocal<object>();
+
+                var sentinel = new object();
+
+                var sem = new SemaphoreSlim(0);
+                Task continuation = sem.WaitAsync(cts.Token).ContinueWith(prev =>
+                {
+                    Assert.Equal(TaskStatus.Canceled, prev.Status);
+                    Assert.NotSame(sentinel, tl.Value);
```

Comment:
Should we add some code to deadlock if it executed synchronously?

Reply:
What would that demonstrate that this doesn't?  I'm happy to improve the validation if there's something missing here.

Reply:
I was just thinking the test code can demonstrate the real problem when executing synchronously. but you are right, will not add any testing value.